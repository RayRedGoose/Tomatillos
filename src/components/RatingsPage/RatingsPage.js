import React, { Component } from 'react';
import { connect } from 'react-redux';
import { NavLink } from 'react-router-dom';
import './RatingsPage.scss'
import SmallMovieCard from '../SmallMovieCard/SmallMovieCard'

export class RatingsPage extends Component {
  createCards = (movies, ratings) => {
    return ratings.map(rating => {
      const movie = movies.find(movie => movie.id === rating.movie_id);
      return (
        <SmallMovieCard
          key={rating.id}
          id={movie.id}
          img={movie.poster_path}
          rate={rating.rating}
          title={movie.title}
          deleteButton={true}/>
      )
    });
  }

  render() {
    let { movies, ratings, user } = this.props;
    const notification = (user)
      ? 'You don\'t have any rating yet!'
      : 'Please, log in to see your ratings!'

    return (
      <div className="ratings">
        <NavLink to='/'>&#8592; back to all movies</NavLink>
        { (user && ratings.length !== 0)
            ? this.createCards(movies, ratings)
            : <h2 className="error">{notification}</h2>
        }
      </div>
    );
  }
}

export const mapStateToProps = ({movies, ratings, user}) => ({
  movies,
  ratings,
  user
})

export default connect(mapStateToProps)(RatingsPage);
