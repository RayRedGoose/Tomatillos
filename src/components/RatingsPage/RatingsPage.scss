.ratings {
  display: flex;
  flex-wrap: wrap;
  height: 80vh;
  overflow: scroll;
  margin: 20px 20px;
  padding: 40px 20px 20px;
  position: relative;
  a {
    color: #fff;
    text-decoration: none;
    cursor: pointer;
    font-size: 18px;
    background: none;
    border: none;
    position: absolute;
    top: 0; left: 0;
    margin: 0;
    margin-left: 20px;
    &:hover {
      transform: scale(1.1);
    }
  }
}
