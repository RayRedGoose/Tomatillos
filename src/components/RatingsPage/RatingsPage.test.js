import React from 'react';
import { RatingsPage, mapStateToProps } from './RatingsPage';
import { shallow } from 'enzyme';

describe("RatingsPage", () => {
  const mockMoviesData =
    [{
      id: 1,
      title: "Jumanji: The Next Level",
      poster_path: 'https://some.url/poster-image',
      backdrop_path: 'https://some.url/bg-image',
      release_date: '2019-12-04',
      overview: 'In Jumanji: The Next Level, the gang is back but the game has changed.',
      average_rating: 4
    }];

  const mockRatingsData =
    [{
      id: 1,
      user_id: 1,
      movie_id: 1,
      rating: 6,
      created_at: "someDate",
      updated_at: "someDate"
    }];

  const mockProps = {
    movies: mockMoviesData,
    ratings: mockRatingsData,
    login: true
  }

  describe("RatingsPage component", () => {
    let page, instance;

    beforeEach(() => {
      page = shallow(<RatingsPage {...mockProps}/>);
      instance = page.instance();
    });

    it("should match snapshop with all data passed in correctly", () => {
      expect(page).toMatchSnapshot();
    });

    it("should match snapshop if there is user logged but no ratings", () => {
      const pageNoRatings = shallow(<RatingsPage {...mockProps} ratings={[]} />);
      expect(pageNoRatings).toMatchSnapshot();
    });

    it("should match snapshop if there is no user logged", () => {
      const pageNoUser = shallow(<RatingsPage {...mockProps} login={false} />)
      expect(pageNoUser).toMatchSnapshot();
    });

    it("should match snapshop if redirect is true", () => {
      page.setState({isRedirected: true})
      expect(page).toMatchSnapshot();
    });

    it("should call createCards after rendering", () => {
      const spy = jest.spyOn(instance, 'createCards').mockImplementation(() => {});
      instance.forceUpdate();
      expect(spy).toHaveBeenCalled();
    });

    it("should call redirect after back button was clicked", () => {
      const spy = jest.spyOn(instance, 'redirect').mockImplementation(() => {});
      instance.forceUpdate();

      page.find('button').simulate('click');

      expect(spy).toHaveBeenCalled();
    });

    it("should change isRedirect state to true when redirect is called", () => {
      expect(page.state('isRedirected')).toEqual(false);

      instance.redirect();

      expect(page.state('isRedirected')).toEqual(true);
    });
  });

  describe("mapStateToProps", () => {
    it("shoud return an object with movies, ratings arrays and login state", () => {
      const mockState = {
        ...mockProps,
        isLoaded: true,
        error: ''
      };

      const expected = {
        ...mockProps
      };

      const mappedProps = mapStateToProps(mockState);

      expect(mappedProps).toEqual(expected);
    });
  });
});
