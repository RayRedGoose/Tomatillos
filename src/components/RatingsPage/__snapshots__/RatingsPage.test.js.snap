// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RatingsPage RatingsPage component should match snapshop if there is no user logged 1`] = `
<div
  className="ratings"
>
  <NavLink
    to="/"
  >
    ← back to all movies
  </NavLink>
  <h2
    className="error"
  >
    Please, log in to see your ratings!
  </h2>
</div>
`;

exports[`RatingsPage RatingsPage component should match snapshop if there is user logged but no ratings 1`] = `
<div
  className="ratings"
>
  <NavLink
    to="/"
  >
    ← back to all movies
  </NavLink>
  <h2
    className="error"
  >
    You don't have any rating yet!
  </h2>
</div>
`;

exports[`RatingsPage RatingsPage component should match snapshop with all data passed in correctly 1`] = `
<div
  className="ratings"
>
  <NavLink
    to="/"
  >
    ← back to all movies
  </NavLink>
  <SmallMovieCard
    deleteButton={true}
    id={1}
    img="https://some.url/poster-image"
    key="1"
    rate={6}
    title="Jumanji: The Next Level"
  />
</div>
`;
